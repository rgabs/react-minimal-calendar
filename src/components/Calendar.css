.overflow-days {
    opacity: 0.3;
}