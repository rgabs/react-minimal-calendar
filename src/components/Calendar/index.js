import React, { PureComponent } from "react";
import moment from "moment";
import uuid from "short-uuid";
import setQueryString from "set-query-string";
import getQueryString from "query-string";
import CalendarControls from "../CalendarControls";
import CalendarPane from "../CalendarPane";

class Calendar extends PureComponent {
  state = {
    momentInstance: moment()
  };

  componentDidMount() {
    const monthIndex = getQueryString.parse(location.search).month;
    if (monthIndex) {
      this.setMonth(monthIndex);
    }
  }

  days = moment.weekdaysShort();
  months = moment.months();

  reset = () => {
    this.setState({ momentInstance: moment() });
    setQueryString({ month: "" });
    this.props.onChange(null);
  };

  years = (() => {
    const years = [];
    for (let i = moment().year() - 20; i < moment().year() + 20; i++) {
      years.push(i);
    }
    return years;
  })();

  getOverflowDays = () => {
    const blanks = [];
    const firstDay = moment(this.state.momentInstance)
      .startOf("month")
      .format("d");
    let daysInPreviousMonth = moment(this.state.momentInstance)
      .subtract(1, "months")
      .daysInMonth();
    for (let i = firstDay - 1; i >= 0; i--) {
      blanks.push(
        <td className="overflow-days" key={uuid.generate()}>
          {daysInPreviousMonth - i}
        </td>
      );
    }
    return blanks;
  };

  isToday = dayOfMonth =>
    moment().isSame(this.state.momentInstance, "year") &&
    moment().isSame(this.state.momentInstance, "month") &&
    dayOfMonth === moment().date();

  getMonthDates = () => {
    const dates = [];
    const totalDaysInMonth = this.state.momentInstance.daysInMonth();
    for (let i = 1; i <= totalDaysInMonth; i++) {
      dates.push(
        <td
          key={uuid.generate()}
          onClick={this.onDateSelect(i)}
          className={this.isToday(i) ? "date-active" : ""}
        >
          {i}
        </td>
      );
    }
    return dates;
  };

  onDateSelect = dayOfMonth => () => {
    const newInstance = moment({ ...this.state.momentInstance }).set(
      "date",
      dayOfMonth
    );
    this.setState({ momentInstance: newInstance });
    this.props.onChange(newInstance);
  };

  groupDaysByWeeks = allDays => {
    const overflowedDays = [...allDays];
    const groupedDays = [];
    const totalRequiredSlots = Math.ceil(allDays.length / 7) * 7;
    for (let j = 1; j <= totalRequiredSlots - allDays.length; j++) {
      overflowedDays.push(
        <td className="overflow-days" key={uuid.generate()}>
          {j}
        </td>
      );
    }
    const numOfWeeks = Math.ceil(overflowedDays.length / 7);
    for (let i = 0; i < numOfWeeks; i++) {
      const daysInWeek = overflowedDays.slice(i * 7, (i + 1) * 7);
      groupedDays.push(<tr key={uuid.generate()}>{daysInWeek}</tr>);
    }
    return groupedDays;
  };

  setMonth = monthIndex => {
    const newInstance = moment({ ...this.state.momentInstance }).set(
      "month",
      monthIndex
    );
    this.setState({ momentInstance: newInstance });
    setQueryString({ month: monthIndex });
  };

  onMonthChange = e => this.setMonth(e.target.value);

  setYear = e => {
    const newInstance = moment({ ...this.state.momentInstance }).set(
      "year",
      e.target.value
    );
    this.setState({ momentInstance: newInstance });
  };

  setPreviousMonth = () => {
    const prevMonthIndex = moment({ ...this.state.momentInstance })
      .subtract(1, "months")
      .month();
    this.setMonth(prevMonthIndex);
  };

  setNextMonth = () => {
    const nextMonthIndex = moment({ ...this.state.momentInstance })
      .add(1, "months")
      .month();
    this.setMonth(nextMonthIndex);
  };

  isNextMonthBtnDisabled = () =>
    moment(this.state.momentInstance).month() === 11 &&
    moment(this.state.momentInstance).year() ===
      this.years[this.years.length - 1];

  isPrevMonthBtnDisabled = () =>
    moment(this.state.momentInstance).month() === 0 &&
    moment(this.state.momentInstance).year() === this.years[0];

  render() {
    const monthDates = this.getMonthDates();
    const overflowDays = this.getOverflowDays(monthDates);
    const slots = this.groupDaysByWeeks([...overflowDays, ...monthDates]);
    return (
      <div>
        <CalendarControls
          isPrevMonthBtnDisabled={this.isPrevMonthBtnDisabled()}
          setPreviousMonth={this.setPreviousMonth}
          month={this.state.momentInstance.month()}
          months={this.months}
          onMonthChange={this.onMonthChange}
          year={this.state.momentInstance.year()}
          onYearChange={this.setYear}
          years={this.years}
          isNextMonthBtnDisabled={this.isNextMonthBtnDisabled()}
          setNextMonth={this.setNextMonth}
          reset={this.reset}
        />
        <CalendarPane daysHeader={this.days} slots={slots} />
      </div>
    );
  }
}

Calendar.propTypes = {
  onChange: () => {}
};

export default Calendar;
