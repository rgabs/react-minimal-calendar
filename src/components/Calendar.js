import React, { PureComponent } from "react";
import moment from "moment";
import uuid from "short-uuid";
import "./Calendar.css";

class Calendar extends PureComponent {
  state = {
    momentInstance: moment()
  };

  days = moment.weekdaysShort();
  months = moment.months();

  years = (() => {
    const years = [];
    for (let i = moment().year() - 20; i < moment().year() + 20; i++) {
      years.push(i);
    }
    return years;
  })();

  getOverflowDays = () => {
    const blanks = [];
    const firstDay = moment(this.state.momentInstance)
      .startOf("month")
      .format("d");
    let daysInPreviousMonth = moment(this.state.momentInstance)
      .subtract(1, "months")
      .daysInMonth();
    for (let i = firstDay - 1; i >= 0; i--) {
      blanks.push(
        <td className="overflow-days" key={uuid.generate()}>
          {daysInPreviousMonth - i}
        </td>
      );
    }
    return blanks;
  };

  getMonthDates = () => {
    const dates = [];
    const isToday = dayOfMonth =>
      moment().isSame(this.state.momentInstance, "year") &&
      moment().isSame(this.state.momentInstance, "month") &&
      dayOfMonth === this.state.momentInstance.date();

    const totalDaysInMonth = this.state.momentInstance.daysInMonth();
    for (let i = 1; i <= totalDaysInMonth; i++) {
      dates.push(
        <td key={uuid.generate()} className={isToday(i) ? "date-active" : ""}>
          {i}
        </td>
      );
    }
    return dates;
  };

  groupDaysByWeeks = allDays => {
    const overflowedDays = [...allDays];
    const totalRequiredSlots = Math.ceil(allDays.length / 7) * 7;
    for (let j = 1; j <= totalRequiredSlots - allDays.length; j++) {
      overflowedDays.push(
        <td className="overflow-days" key={uuid.generate()}>
          {j}
        </td>
      );
    }
    const numOfWeeks = Math.ceil(overflowedDays.length / 7);
    const groupedDays = [];
    for (let i = 0; i < numOfWeeks; i++) {
      const daysInWeek = overflowedDays.slice(i * 7, (i + 1) * 7);
      groupedDays.push(<tr key={uuid.generate()}>{daysInWeek}</tr>);
    }
    return groupedDays;
  };

  setMonth = e => {
    const newInstance = moment({ ...this.state.momentInstance }).set(
      "month",
      e.target.value
    );
    this.setState({ momentInstance: newInstance });
  };

  setYear = e => {
    const newInstance = moment({ ...this.state.momentInstance }).set(
      "year",
      e.target.value
    );
    this.setState({ momentInstance: newInstance });
  };

  setPreviousMonth = () => {
    const newInstance = moment({ ...this.state.momentInstance }).subtract(
      1,
      "months"
    );
    this.setState({ momentInstance: newInstance });
  };

  setNextMonth = () => {
    const newInstance = moment({ ...this.state.momentInstance }).add(
      1,
      "months"
    );
    this.setState({ momentInstance: newInstance });
  };

  isNextMonthBtnDisabled = () =>
    moment(this.state.momentInstance).month() === 11 &&
    moment(this.state.momentInstance).year() ===
      this.years[this.years.length - 1];

  isPrevMonthBtnDisabled = () =>
    moment(this.state.momentInstance).month() === 0 &&
    moment(this.state.momentInstance).year() === this.years[0];

  render() {
    const daysHeader = this.days.map((day, i) => <th key={i}>{day}</th>);
    const monthDates = this.getMonthDates();
    const overflowDays = this.getOverflowDays(monthDates);
    const slots = this.groupDaysByWeeks([...overflowDays, ...monthDates]);
    return (
      <div>
        <button
          disabled={this.isPrevMonthBtnDisabled()}
          onClick={this.setPreviousMonth}
        >
          {"<"}
        </button>
        <select
          value={this.state.momentInstance.month()}
          onChange={this.setMonth}
        >
          {this.months.map((month, i) => (
            <option value={i} key={i}>
              {month}
            </option>
          ))}
        </select>
        <select
          value={this.state.momentInstance.year()}
          onChange={this.setYear}
        >
          {this.years.map((year, i) => (
            <option value={year} key={i}>
              {year}
            </option>
          ))}
        </select>
        <button
          disabled={this.isNextMonthBtnDisabled()}
          onClick={this.setNextMonth}
        >
          {">"}
        </button>
        <table>
          <thead className="calendar-header">
            <tr className="calendar-row">{daysHeader}</tr>
          </thead>
          <tbody>{slots}</tbody>
        </table>
      </div>
    );
  }
}

export default Calendar;
