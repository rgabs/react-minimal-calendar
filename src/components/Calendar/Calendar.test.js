import React from 'react';
import Calendar from './Calendar';
import { shallow } from 'enzyme';
import uuid from 'short-uuid';
import renderer from 'react-test-renderer';
const snapshotDiff = require('snapshot-diff');

describe('Calendar component', () => {
    let wrapper, componentInstance;
    beforeEach(() => {
        wrapper = shallow(<Calendar  />);
        componentInstance = wrapper.instance();
    })

    it('renders correctly', () => {
        const tree = renderer.create(
            <Calendar />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('groupDaysByWeeks: 21 days', () => {
        const input = [
            <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>,
            <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>,
            <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>,
        ];
        const output = [
            <tr><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td></tr>,
            <tr><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td></tr>,
            <tr><td>3</td><td>3</td><td>3</td><td>3</td><td>3</td><td>3</td><td>3</td></tr>
        ];
        expect(snapshotDiff(componentInstance.groupDaysByWeeks(input), output)).toBe(`Snapshot Diff:
Compared values have no visual difference.`);
    })

    it('groupDaysByWeeks: 18 days', () => {
        const input = [
            <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, <td key={uuid.generate()}>1</td>, 
            <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>, <td key={uuid.generate()}>2</td>,
            <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>, <td key={uuid.generate()}>3</td>,
        ];
        const output = [
            <tr><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>2</td></tr>,
            <tr><td>2</td><td>2</td><td>2</td><td>2</td><td>2</td><td>3</td><td>3</td></tr>,
            <tr><td>3</td><td>3</td><td>3</td><td>3</td></tr>
        ];
        expect(snapshotDiff(componentInstance.groupDaysByWeeks(input), output)).toBe(`Snapshot Diff:
Compared values have no visual difference.`);
    })
});