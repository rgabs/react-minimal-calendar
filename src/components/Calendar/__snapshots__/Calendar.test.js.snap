// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component renders correctly 1`] = `
<div>
  <select
    onChange={[Function]}
    value={0}
  >
    <option
      value={0}
    >
      January
    </option>
    <option
      value={1}
    >
      February
    </option>
    <option
      value={2}
    >
      March
    </option>
    <option
      value={3}
    >
      April
    </option>
    <option
      value={4}
    >
      May
    </option>
    <option
      value={5}
    >
      June
    </option>
    <option
      value={6}
    >
      July
    </option>
    <option
      value={7}
    >
      August
    </option>
    <option
      value={8}
    >
      September
    </option>
    <option
      value={9}
    >
      October
    </option>
    <option
      value={10}
    >
      November
    </option>
    <option
      value={11}
    >
      December
    </option>
  </select>
  <select
    onChange={[Function]}
    value={2019}
  >
    <option
      value={1999}
    >
      1999
    </option>
    <option
      value={2000}
    >
      2000
    </option>
    <option
      value={2001}
    >
      2001
    </option>
    <option
      value={2002}
    >
      2002
    </option>
    <option
      value={2003}
    >
      2003
    </option>
    <option
      value={2004}
    >
      2004
    </option>
    <option
      value={2005}
    >
      2005
    </option>
    <option
      value={2006}
    >
      2006
    </option>
    <option
      value={2007}
    >
      2007
    </option>
    <option
      value={2008}
    >
      2008
    </option>
    <option
      value={2009}
    >
      2009
    </option>
    <option
      value={2010}
    >
      2010
    </option>
    <option
      value={2011}
    >
      2011
    </option>
    <option
      value={2012}
    >
      2012
    </option>
    <option
      value={2013}
    >
      2013
    </option>
    <option
      value={2014}
    >
      2014
    </option>
    <option
      value={2015}
    >
      2015
    </option>
    <option
      value={2016}
    >
      2016
    </option>
    <option
      value={2017}
    >
      2017
    </option>
    <option
      value={2018}
    >
      2018
    </option>
    <option
      value={2019}
    >
      2019
    </option>
    <option
      value={2020}
    >
      2020
    </option>
    <option
      value={2021}
    >
      2021
    </option>
    <option
      value={2022}
    >
      2022
    </option>
    <option
      value={2023}
    >
      2023
    </option>
    <option
      value={2024}
    >
      2024
    </option>
    <option
      value={2025}
    >
      2025
    </option>
    <option
      value={2026}
    >
      2026
    </option>
    <option
      value={2027}
    >
      2027
    </option>
    <option
      value={2028}
    >
      2028
    </option>
    <option
      value={2029}
    >
      2029
    </option>
    <option
      value={2030}
    >
      2030
    </option>
    <option
      value={2031}
    >
      2031
    </option>
    <option
      value={2032}
    >
      2032
    </option>
    <option
      value={2033}
    >
      2033
    </option>
    <option
      value={2034}
    >
      2034
    </option>
    <option
      value={2035}
    >
      2035
    </option>
    <option
      value={2036}
    >
      2036
    </option>
    <option
      value={2037}
    >
      2037
    </option>
    <option
      value={2038}
    >
      2038
    </option>
  </select>
  <table>
    <thead
      className="calendar-header"
    >
      <tr
        className="calendar-row"
      >
        <th>
          Sun
        </th>
        <th>
          Mon
        </th>
        <th>
          Tue
        </th>
        <th>
          Wed
        </th>
        <th>
          Thu
        </th>
        <th>
          Fri
        </th>
        <th>
          Sat
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td />
        <td />
        <td
          className=""
        >
          1
        </td>
        <td
          className=""
        >
          2
        </td>
        <td
          className=""
        >
          3
        </td>
        <td
          className=""
        >
          4
        </td>
        <td
          className=""
        >
          5
        </td>
      </tr>
      <tr>
        <td
          className=""
        >
          6
        </td>
        <td
          className=""
        >
          7
        </td>
        <td
          className=""
        >
          8
        </td>
        <td
          className=""
        >
          9
        </td>
        <td
          className=""
        >
          10
        </td>
        <td
          className=""
        >
          11
        </td>
        <td
          className=""
        >
          12
        </td>
      </tr>
      <tr>
        <td
          className=""
        >
          13
        </td>
        <td
          className=""
        >
          14
        </td>
        <td
          className=""
        >
          15
        </td>
        <td
          className=""
        >
          16
        </td>
        <td
          className=""
        >
          17
        </td>
        <td
          className=""
        >
          18
        </td>
        <td
          className=""
        >
          19
        </td>
      </tr>
      <tr>
        <td
          className=""
        >
          20
        </td>
        <td
          className=""
        >
          21
        </td>
        <td
          className=""
        >
          22
        </td>
        <td
          className="active"
        >
          23
        </td>
        <td
          className=""
        >
          24
        </td>
        <td
          className=""
        >
          25
        </td>
        <td
          className=""
        >
          26
        </td>
      </tr>
      <tr>
        <td
          className=""
        >
          27
        </td>
        <td
          className=""
        >
          28
        </td>
        <td
          className=""
        >
          29
        </td>
        <td
          className=""
        >
          30
        </td>
        <td
          className=""
        >
          31
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
